// anilist.js/dist/client/utilities.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const node_fetch_1 = tslib_1.__importDefault(require("node-fetch"));
class Utilities {
    APIRequest(query, variables, client) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const url = 'https://graphql.anilist.co', options = {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    Accept: 'application/json',
                },
                body: JSON.stringify({ query, variables }),
            };
            if (client.token)
                Object.assign(options.headers, { Authorization: 'Bearer ' + client.token });
            const req = yield node_fetch_1.default(url, options);
            const json = yield req.json();
            if (json.errors)
                throw new Error(JSON.stringify(json.errors));
            return (json.data.Character || json.data.Media || json.data.Page || json.data.User || json.data.Studio || json.data.Staff || json.data.Viewer);
        });
    }
}
exports.default = Utilities;
